sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -64,21 +64,18 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
                 throw new InvalidOperationException("Bearer token authentication is not permitted for non TLS protected (https) endpoints.");
             }
 
-            if (DateTimeOffset.UtcNow >= _refreshOn)
+            (string? headerValue, bool refreshTokenInBackground) = await _accessTokenCache.GetHeaderValueAsync(async, message.CancellationToken);
+            if (headerValue == null)
             {
-                AccessToken token = async ?
-                        await _credential.GetTokenAsync(new TokenRequestContext(_scopes, message.Request.ClientRequestId), message.CancellationToken).ConfigureAwait(false) :
-                        _credential.GetToken(new TokenRequestContext(_scopes, message.Request.ClientRequestId), message.CancellationToken);
-
-                _headerValue = "Bearer " + token.Token;
-                _refreshOn = token.ExpiresOn - TimeSpan.FromMinutes(2);
+                headerValue = await GetHeaderValueFromCredentialAsync(message, async);
             }
-
-            if (_headerValue != null)
+            else if (refreshTokenInBackground)
             {
-                message.Request.SetHeader(HttpHeader.Names.Authorization, _headerValue);
+                _ = Task.Run(() => GetHeaderValueFromCredentialAsync(message, async));
```

Comment:
Can we handle this in GetHeaderValueAsync?

Reply:
Not really. It either requires to pass a callback or a direct reference to the `TokenCredential`. I have renamed `refreshTokenInBackground` into `accessTokenAboutToExpire`.

Reply:
Resolved with new implementation